test/config-luatest/replicaset_test.lua

```diff
@@ -0,0 +1,64 @@
+local t = require('luatest')
+local cbuilder = require('test.config-luatest.cbuilder')
+local replicaset = require('test.config-luatest.replicaset')
+
+local g = t.group()
+
+g.before_all(replicaset.init)
+g.after_each(replicaset.clean)
+g.after_all(replicaset.clean)
+
+-- {{{ General purpose helpers
+
+local function toline(s)
+    return s:gsub('\n', ''):gsub(' +', ' '):strip()
+end
+
+-- }}} General purpose helpers
+
+-- {{{ Failure cases
+
+-- Verify that replication.election_mode can't be set other than
+-- null or "off" if replication.failover is not "election".
+--
+-- The whole replicaset refuses to start if this misconfiguration
+-- is found.
+g.test_election_mode_without_failover_election = function(g)
+    local error_t = toline([[
+        replication.election_mode = %q is set for instance "instance-004" of
+        replicaset "replicaset-001" of group "group-001", but this option is
+        only applicable if replication.failover = "election"; the replicaset is
+        configured with replication.failover = %q; if this particular instance
+        requires its own election mode, for example, if it is an anonymous
+        replica, consider configuring the election mode specifically for this
+        particular instance
+    ]])
+
```

Comment:
Nit: we should probably add a test, that with `election_mode = "off"` everything is ok

Reply:
It is a bit cumbersome to test all the possible compositions of these options, but since I actually made a mistake here, it definitely worth it. Thank you for noticing! Will do.

Reply:
As result of the test rewriting I need the following changes in the configuration builder.

```diff
diff --git a/test/config-luatest/cbuilder.lua b/test/config-luatest/cbuilder.lua
index 5c2d73eb65..e33925d72a 100644
--- a/test/config-luatest/cbuilder.lua
+++ b/test/config-luatest/cbuilder.lua
@@ -69,6 +69,24 @@ local base_config = {
             peer = 'replicator@',
         },
     },
+    replication = {
+        -- The default value is 1 second. It is good for a real
+        -- usage, but often suboptimal for testing purposes.
+        --
+        -- If an instance can't connect to another instance (say,
+        -- because it is not started yet), it retries the attempt
+        -- after so called 'replication interval', which is equal
+        -- to replication timeout.
+        --
+        -- One second waiting means one more second for a test
+        -- case and, if there are many test cases with a
+        -- replicaset construction, it affects the test timing a
+        -- lot.
+        --
+        -- replication.timeout = 0.1 second reduces the timing
+        -- by half for my test.
+        timeout = 0.1,
+    },
 }
 
 -- Set an option for replicaset-001.
@@ -85,7 +103,13 @@ local function cbuilder_set_replicaset_option(self, path, value)
     if value == nil then
         local cur = self._config
         for i = 1, #path - 1 do
-            cur = cur[path[i]]
+            -- Create missed fields.
+            local component = path[i]
+            if cur[component] == nil then
+                cur[component] = {}
+            end
+
+            cur = cur[component]
         end
         cur[path[#path]] = value
         return self
```

I'll apply them right into the commit, where it is added.

I'll also add the changes for the replicaset testing helper commit:

```diff
diff --git a/test/config-luatest/replicaset.lua b/test/config-luatest/replicaset.lua
index 05e892a313..301cbc5395 100644
--- a/test/config-luatest/replicaset.lua
+++ b/test/config-luatest/replicaset.lua
@@ -20,16 +20,16 @@ local t = require('luatest')
 local treegen = require('test.treegen')
 local justrun = require('test.justrun')
 
-local function replicaset_init(g)
+local function init(g)
     treegen.init(g)
 end
 
 -- Stop all the managed instances using <server>:drop().
-local function replicaset_drop(_g)
+local function drop(_g)
     -- Not implemented yet.
 end
 
-local function replicaset_clean(g)
+local function clean(g)
     treegen.clean(g)
 end
 
@@ -55,7 +55,7 @@ end
 -- Starts a all instance of a replicaset from the given config and
 -- ensure that all the instances fails to start and reports the
 -- given error message.
-local function replicaset_startup_error(g, config, exp_err)
+local function startup_error(g, config, exp_err)
     -- Prepare a temporary directory and write a configuration
     -- file.
     local dir = treegen.prepare_directory(g, {}, {})
@@ -81,8 +81,8 @@ end
 -- }}} Replicaset that can't start
 
 return {
-    init = replicaset_init,
-    drop = replicaset_drop,
-    clean = replicaset_clean,
-    startup_error = replicaset_startup_error,
+    init = init,
+    drop = drop,
+    clean = clean,
+    startup_error = startup_error,
 }
```

And I'll add one more commit with replicaset management utils.

Next, I'm going to completely rewrite the test and rename it to `failover_and_election_mode_test.lua`.